src/coprocessor/expression/mod.rs

```diff
@@ -0,0 +1,292 @@
+        }
+        Codec(err: util::codec::Error) {
+            from()
+                description("codec error")
+                display("codec error: {}", err)
+                cause(err)
+        }
+        Other(desc: &'static str) {
+            description(desc)
+                display("error {}", desc)
+        }
+    }
+}
+
+pub type Result<T> = ::std::result::Result<T, Error>;
+
+
+/// Flags are used by `SelectRequest.flags` to handle execution mode, like how to handle
+/// truncate error.
+/// `FLAG_IGNORE_TRUNCATE` indicates if truncate error should be ignored.
+/// Read-only statements should ignore truncate error, write statements should not ignore
+/// truncate error.
+pub const FLAG_IGNORE_TRUNCATE: u64 = 1;
+/// `FLAG_TRUNCATE_AS_WARNING` indicates if truncate error should be returned as warning.
+/// This flag only matters if `FLAG_IGNORE_TRUNCATE` is not set, in strict sql mode, truncate error
+/// should be returned as error, in non-strict sql mode, truncate error should be saved as warning.
+pub const FLAG_TRUNCATE_AS_WARNING: u64 = 1 << 1;
+
+/// StatementContext contains variables for a statement
+/// It should be reset before executing a statement
+pub struct StatementContext {
+    ignore_truncate: bool,
+    truncate_as_warning: bool,
+    timezone: FixedOffset,
+}
+
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum TypeClass {
+    String = 0,
+    Real = 1,
+    Int = 2,
+    Decimal = 4,
+    Time = 8,
+}
+
+
+/// Expression represents all scalar expression in SQL.
+#[derive(Debug, Clone, PartialEq)]
+pub struct Expression {
+    pub expr: ExprKind,
+    pub ret_type: TypeClass,
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum ExprKind {
+    Constant(Datum),
+    ScalarFn(BuiltinFn),
+    ColumnRef(usize),
+}
+
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum UnOp {
+    Not,
+    Neg,
+    BitNeg,
+    // ISNULL(expr)
+    IsNull,
+    // IS boolean_value
+    IsTrue,
+    IsFalse,
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum BinOp {
+    // arithmetical
+    Add,
+    Minus,
+    Mul,
+    Div,
+    Rem,
+    IntDiv,
+    // logical
+    And,
+    Or,
+    Xor,
+    // bitwise
+    BitAnd,
+    BitOr,
+    BitXor,
+    Shl,
+    Shr,
+    // compare
+    Lt,
+    Le,
+    Eq,
+    Ne,
+    Ge,
+    Gt,
+    NullEq,
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct CaseArm {
+    pub condition: Option<Box<Expression>>,
+    pub result: Box<Expression>,
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum BuiltinFn {
+    // ## operators
+    Unary(UnOp, Box<Expression>),
+    Binary(BinOp, Box<Expression>, Box<Expression>),
+    // COALESCE(value,...)
+    Coalesce(Vec<Expr>),
+    // GREATEST(value1,value2,...)
+    Greatest(Vec<Expr>),
+    // expr IN (value,...)
+    In(Box<Expression>, Vec<Expr>),
+    // INTERVAL(N,N1,N2,N3,...)
+    Interval(Box<Expression>, Vec<Expr>),
+    // LEAST(value1,value2,...)
+    Least(Vec<Expr>),
```

Comment:
It confused me,  why we need to list operators? In my view it would be more clear if we implement functions  like  `eval`,`eval_int`,`eval_real`  for `BuiltinFun` directly?